crates/bevy_reflect/src/impls/bevy_ecs.rs

```diff
@@ -1,210 +1,81 @@
 use crate::{FromType, Reflect};
 use bevy_ecs::{
-    Archetype, Component, Entity, EntityMap, FromResources, MapEntities, MapEntitiesError,
-    Resources, World,
+    component::Component,
+    entity::{Entity, EntityMap, MapEntities, MapEntitiesError},
+    world::{FromWorld, World},
 };
-use std::marker::PhantomData;
 
 #[derive(Clone)]
 pub struct ReflectComponent {
-    add_component: fn(&mut World, resources: &Resources, Entity, &dyn Reflect),
+    add_component_from_source: fn(&mut World, Entity, &dyn Reflect),
     apply_component: fn(&mut World, Entity, &dyn Reflect),
-    reflect_component: unsafe fn(&Archetype, usize) -> &dyn Reflect,
-    reflect_component_mut: unsafe fn(&Archetype, usize) -> &mut dyn Reflect,
```

Comment:
Why did you remove this? It could be useful I think.

Reply:
+1, this is very useful for `bevy-inspector-egui`

Reply:
We still have `reflect_component` (it just got renamed and now returns an Option instead of assuming valid entity indices). I removed the mutable variant because it requires mutably deref-ing `Mut<T>` (if we want to use the equivalent `EntityRef::get_unchecked_mut` method), which triggers change detection. It wasn't used anywhere internally, so I figured I'd leave it out until we had a use case that forced us to make a call on the change detection.

Given that bevy-inspector-egui uses it currently, I'm down to make a call here. I'm thinking that this _should_ trigger change detection and callers should only call `reflect_component_mut` if they intend to actually mutate the component.

Reply:
Ah wait I just noticed we can't use `get_unchecked_mut` because the tracking pointer owns the reference. That doesn't change the conversation, we just need a custom impl.

Reply:
Hmm `Mut<Reflect>` is also an option. That might actually be the right call.

Reply:
Just re-added reflect_component_mut by adding a new ReflectMut tracking pointer for reflected components. This required moving the bevy_ecs reflect impls into bevy_ecs (which was a long time coming anyway).